src/hist/plot.py

```diff
@@ -210,185 +220,386 @@ def plot2d_full(
+        # bottom can't be nan
+        bar_bottom[np.isnan(bar_bottom)] = 0
+        bar_height = bar_top - bar_bottom
+
+        ax.scatter(x_values, ratio, color="black")
+        ax.bar(
+            x_values,
+            height=bar_height,
+            width=bar_width,
+            bottom=bar_bottom,
+            fill=False,
+            linewidth=0,
+            edgecolor="gray",
+            hatch=3 * "/",
+        )
+
+    ratio_ylim = kwargs.pop("ylim", None)
+    if ratio_ylim is None:
+        # plot centered around central value with a scaled view range
+        # the value _with_ the uncertainty in view is important so base
+        # view range on extrema of value +/- uncertainty
+        valid_ratios_idx = np.where(np.isnan(ratio) == False)  # noqa: E712
+        valid_ratios = ratio[valid_ratios_idx]
+        extrema = np.array(
+            [
+                valid_ratios - ratio_uncert[0][valid_ratios_idx],
+                valid_ratios + ratio_uncert[1][valid_ratios_idx],
+            ]
+        )
+        max_delta = np.max(np.abs(extrema - central_value))
+        ratio_extrema = np.abs(max_delta + central_value)
+
+        _alpha = 2.0
+        scaled_offset = max_delta + (max_delta / (_alpha * ratio_extrema))
+        ratio_ylim = [central_value - scaled_offset, central_value + scaled_offset]
+
+    ax.set_xlim(left_edge, right_edge)
+    ax.set_ylim(bottom=ratio_ylim[0], top=ratio_ylim[1])
+
+    ax.set_xlabel(_hist.axes[0].label)
+    ax.set_ylabel(kwargs.pop("ylabel", "Ratio"))
+
+    return ax
+
+
 def plot_pull(
+    _hist: hist.BaseHist,
+    pulls: np.ndarray,
+    ax: matplotlib.axes.Axes,
+    bar_kwargs: Dict[str, Any],
+    pp_kwargs: Dict[str, Any],
+) -> matplotlib.axes.Axes:
+    """
+    Plot a pull plot on the given axes
+    """
+    x_values = _hist.axes[0].centers
+    left_edge = _hist.axes.edges[0][0]
+    right_edge = _hist.axes.edges[-1][-1]
+
+    # Pull: plot the pulls using Matplotlib bar method
+    width = (right_edge - left_edge) / len(pulls)
+    ax.bar(x_values, pulls, width=width, **bar_kwargs)
+
+    pp_num = pp_kwargs.pop("num", 5)
+    patch_height = max(np.abs(pulls)) / pp_num
+    patch_width = width * len(pulls)
+    for i in range(pp_num):
+        # gradient color patches
+        if "alpha" in pp_kwargs:
+            pp_kwargs["alpha"] *= np.power(0.618, i)
+        else:
+            pp_kwargs["alpha"] = 0.5 * np.power(0.618, i)
+
+        upRect_startpoint = (left_edge, i * patch_height)
+        upRect = patches.Rectangle(
+            upRect_startpoint, patch_width, patch_height, **pp_kwargs
+        )
+        ax.add_patch(upRect)
+        downRect_startpoint = (left_edge, -(i + 1) * patch_height)
+        downRect = patches.Rectangle(
+            downRect_startpoint, patch_width, patch_height, **pp_kwargs
+        )
+        ax.add_patch(downRect)
+
+    ax.set_xlim(left_edge, right_edge)
+
+    ax.set_xlabel(_hist.axes[0].label)
+    ax.set_ylabel("Pull")
+
+    return ax
```

Comment:
Note to self: This should also be returning the axis artists like the above requests.